src/gems/simulation/extra_output.py

```diff
@@ -0,0 +1,155 @@
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict
+
+from gems.expression import evaluate
+from gems.expression.evaluate import EvaluationError, ValueProvider
+from gems.expression.expression import ExpressionNode
+from gems.study.data import ComponentParameterIndex, TimeScenarioIndex
+
+
+@dataclass
+class ExtraOutput:
+    """Stores evaluated outputs (from ExpressionNodes), not solver variables."""
+
+    name: str
+    values: Dict[TimeScenarioIndex, float] = field(default_factory=dict)
+
+    def set(self, t: int, s: int, value: float) -> None:
+        self.values[TimeScenarioIndex(t, s)] = value
+
+    def get(self, t: int, s: int) -> float | None:
+        return self.values.get(TimeScenarioIndex(t, s))
+
+
+def evaluate_all_extra_outputs(
+    problem: Any, component_lookup: Callable[[str], Any]
+) -> Dict[str, Dict[str, ExtraOutput]]:
+    """
+    Evaluate all model-defined extra outputs for every component in the network.
+    """
+    if problem is None or getattr(problem, "context", None) is None:
+        return {}
```

Comment:
Applying previous remark, this could simply become
```suggestion
    if problem is None:
        return {}
```
As otherwise, `problem` is necessarily an `OptimizationProblem` and has all required attributes